File: moose-examples/snippets/vclamp.py
# vclamp.py --- 
# 
# Filename: vclamp.py
# Description: 
# Author:Subhasis Ray 
# Maintainer: 
# Created: Sat Feb  2 19:16:54 2013 (+0530)
# Version: 
# Last-Updated: Tue Jun 11 17:35:20 2013 (+0530)
#           By: subha
#     Update #: 178
# URL: 
# Keywords: 
# Compatibility: 
# 
# 

# Commentary: 
# 
# 
# 
# 

# Change log:
# 
# 
# 
# 
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 3, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING.  If not, write to
# the Free Software Foundation, Inc., 51 Franklin Street, Fifth
# Floor, Boston, MA 02110-1301, USA.
# 
# 

# Code:

import sys
sys.path.append('../../python')
import moose
sys.path.append('../squid')
from squid import SquidAxon
from pylab import *

def vclamp_demo(simtime=50.0, dt=1e-2):
    ## It is good practice to modularize test elements inside a
    ## container
    container = moose.Neutral('/vClampDemo')
    ## Create a compartment with properties of a squid giant axon
    comp = SquidAxon('/vClampDemo/axon')
    # Create and setup the voltage clamp object
    clamp = moose.VClamp('/vClampDemo/vclamp')
    ## The defaults should work fine
    # clamp.mode = 2
    # clamp.tau = 10*dt
    # clamp.ti = dt
    # clamp.td = 0
    # clamp.gain = comp.Cm / dt
    ## Setup command voltage time course
    command = moose.PulseGen('/vClampDemo/command')
    command.delay[0] = 10.0
    command.width[0] = 20.0
    command.level[0] = 50.0
    command.delay[1] = 1e9
    moose.connect(command, 'output', clamp, 'commandIn')
    ## Connect the Voltage Clamp to the compartemnt
    moose.connect(clamp, 'currentOut', comp, 'injectMsg')
    moose.connect(comp, 'VmOut', clamp, 'sensedIn')
    ## setup stimulus recroding - this is the command pulse
    stimtab = moose.Table('/vClampDemo/vclamp_command')
    moose.connect(stimtab, 'requestOut', command, 'getOutputValue')
    ## Set up Vm recording
    vmtab = moose.Table('/vClampDemo/vclamp_Vm')
    moose.connect(vmtab, 'requestOut', comp, 'getVm')
    ## setup command potential recording - this is the filtered input
    ## to PID controller
    commandtab = moose.Table('/vClampDemo/vclamp_filteredcommand')
    moose.connect(commandtab, 'requestOut', clamp, 'getCommand')
    ## setup current recording
    Imtab = moose.Table('/vClampDemo/vclamp_inject')
    moose.connect(Imtab, 'requestOut', clamp, 'getCurrent')
    # Scheduling
    moose.setClock(0, dt)
    moose.setClock(1, dt)    
    moose.setClock(2, dt)
    moose.setClock(3, dt)
    moose.useClock(0, '%s/##[TYPE=Compartment]' % (container.path), 'init')
    moose.useClock(0, '%s/##[TYPE=PulseGen]' % (container.path), 'process')
    moose.useClock(1, '%s/##[TYPE=Compartment]' % (container.path), 'process')
    moose.useClock(2, '%s/##[TYPE=HHChannel]' % (container.path), 'process')
    moose.useClock(2, '%s/##[TYPE=VClamp]' % (container.path), 'process')
    moose.useClock(3, '%s/##[TYPE=Table]' % (container.path), 'process')
    moose.reinit()
    print 'RC filter in VClamp:: tau:', clamp.tau
    print 'PID controller in VClamp:: ti:', clamp.ti, 'td:', clamp.td, 'gain:', clamp.gain
    moose.start(simtime)
    print 'Finished simulation for %g seconds' % (simtime)
    tseries = linspace(0, simtime, len(vmtab.vector))
    subplot(211)
    title('Membrane potential and clamp voltage')
    plot(tseries, vmtab.vector, 'g-', label='Vm (mV)')
    plot(tseries, commandtab.vector, 'b-', label='Filtered command (mV)')
    plot(tseries, stimtab.vector, 'r-', label='Command (mV)')
    xlabel('Time (ms)')
    ylabel('Voltage (mV)')
    legend()
    # print len(commandtab.vector)
    subplot(212)
    title('Current through clamp circuit')
    # plot(tseries, stimtab.vector, label='stimulus (uA)')
    plot(tseries, Imtab.vector, label='injected current (uA)')
    xlabel('Time (ms)')
    ylabel('Current (uA)')
    legend()
    show()

if __name__ == '__main__':
    vclamp_demo()
    

# 
# vclamp.py ends here

File: moose-core/biophysics/VClamp.cpp
// VClamp.cpp ---
//
// Filename: VClamp.cpp
// Description:
// Author:
// Maintainer:
// Created: Fri Feb  1 19:30:45 2013 (+0530)
// Version:
// Last-Updated: Tue Jun 11 17:33:16 2013 (+0530)
//           By: subha
//     Update #: 297
// URL:
// Keywords:
// Compatibility:
//
//

// Commentary:
//
// Implementation of Voltage Clamp
//
//

// Change log:
//
//
//
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public License as
// published by the Free Software Foundation; either version 3, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this program; see the file COPYING.  If not, write to
// the Free Software Foundation, Inc., 51 Franklin Street, Fifth
// Floor, Boston, MA 02110-1301, USA.
//
//

// Code:

#include "../basecode/header.h"
#include "../basecode/Dinfo.h"
#include "VClamp.h"

using namespace moose;

const unsigned int VClamp::DERIVATIVE_ON_PV = 1;
const unsigned int VClamp::PROPORTIONAL_ON_PV = 2;


SrcFinfo1< double >* VClamp::currentOut()
{
    static SrcFinfo1< double > currentOut("currentOut",
                                          "Sends out current output of the clamping circuit. This should be"
                                          " connected to the `injectMsg` field of a compartment to voltage clamp"
                                          " it.");
    return &currentOut;
}


const Cinfo * VClamp::initCinfo()
{
    static DestFinfo process("process",
                             "Handles 'process' call on each time step.",
                             new ProcOpFunc< VClamp >( &VClamp::process));
    static DestFinfo reinit("reinit",
                            "Handles 'reinit' call",
                            new ProcOpFunc< VClamp >( &VClamp::reinit));
    static Finfo * processShared[] = {
        &process,
        &reinit
    };

    static SharedFinfo proc("proc",
                            "Shared message to receive Process messages from the scheduler",
                            processShared, sizeof(processShared) / sizeof(Finfo*));

    static ValueFinfo< VClamp, double> command("command",
                                               "Command input received by the clamp circuit.",
                                               &VClamp::setCommand,
                                               &VClamp::getCommand);
    static ValueFinfo< VClamp, unsigned int> mode("mode",
                                                  "Working mode of the PID controller.\n"
                                                  "\n"
                                                  "   mode = 0, standard PID with proportional, integral and derivative"
                                                  " all acting on the error.\n"
                                                  "\n"
                                                  "   mode = 1, derivative action based on command input\n"
                                                  "\n"
                                                  "   mode = 2, proportional action and derivative action are based on"
                                                  " command input.",
                                                  &VClamp::setMode,
                                                  &VClamp::getMode);
    static ValueFinfo< VClamp, double> ti("ti",
                                          "Integration time of the PID controller. Defaults to 1e9, i.e. integral"
                                          " action is negligibly small.",
                                          &VClamp::setTi,
                                          &VClamp::getTi);

    static ValueFinfo< VClamp, double> td("td",
                                          "Derivative time of the PID controller. This defaults to 0,"
                                          "i.e. derivative action is unused.",
                                          &VClamp::setTd,
                                          &VClamp::getTd);
    static ValueFinfo< VClamp, double> tau("tau",
                                          "Time constant of the lowpass filter at input of the PID"
                                           " controller. This smooths out abrupt changes in the input. Set it to "
                                           " 5 * dt or more to avoid overshoots.",
                                          &VClamp::setTau,
                                          &VClamp::getTau);
    static ValueFinfo< VClamp, double> gain("gain",
                                          "Proportional gain of the PID controller.",
                                          &VClamp::setGain,
                                          &VClamp::getGain);
    static ReadOnlyValueFinfo< VClamp, double> current("current",
                                            "The amount of current injected by the clamp into the membrane.",
                                            &VClamp::getCurrent);


    static ReadOnlyValueFinfo<VClamp, double> sensed("sensed",
                                                     "Membrane potential read from compartment.",
                                                     &VClamp::getVin);
    static DestFinfo sensedIn("sensedIn",
                              "The `VmOut` message of the Compartment object should be connected"
                              " here.",
                              new OpFunc1<VClamp, double>( &VClamp::setVin));

    static DestFinfo commandIn("commandIn",
                              "  The command voltage source should be connected to this.",
                              new OpFunc1<VClamp, double>( &VClamp::setCommand));

    static Finfo* vclampFinfos[] = {
        currentOut(),
        &command,
        &current,
        &sensed,
        &mode,
        &ti,
        &td,
        &tau,
        &gain,
        &sensedIn,
        &commandIn,
        &proc
    };

    static string doc[] = {
        "Name", "VClamp",
        "Author", "Subhasis Ray",
        "Description", "Voltage clamp object for holding neuronal compartments at a specific"
        " voltage.\n"
        "\n"
        "This implementation uses a builtin RC circuit to filter the "
        " command input and then use a PID to bring the sensed voltage (Vm from"
        " compartment) to the filtered command potential.\n"
        "\n"
        "Usage: Connect the `currentOut` source of VClamp to `injectMsg`"
        " dest of Compartment. Connect the `VmOut` source of Compartment to"
        " `set_sensed` dest of VClamp. Either set `command` field to a"
        " fixed value, or connect an appropriate source of command potential"
        " (like the `outputOut` message of an appropriately configured"
        " PulseGen) to `set_command` dest.\n"
        "The default settings for the RC filter and PID controller should be"
        " fine. For step change in command voltage, good defaults with"
        "integration time step dt are as follows:\n"
        "    time constant of RC filter, tau = 5 * dt\n"
        "    proportional gain of PID, gain = Cm/dt where Cm is the membrane"
        "    capacitance of the compartment\n"
        "    integration time of PID, ti = dt\n"
        "    derivative time  of PID, td = 0\n"
        "\n",
    };

    static Dinfo< VClamp > dinfo;
    static Cinfo vclampCinfo(
        "VClamp",
        Neutral::initCinfo(),
        vclampFinfos,
        sizeof(vclampFinfos) / sizeof(Finfo*),
		&dinfo,
        doc,
        sizeof(doc)/sizeof(string));

    return &vclampCinfo;
}

static const Cinfo * vclampCinfo = VClamp::initCinfo();

VClamp::VClamp(): vIn_(0.0), command_(0.0), current_(0.0), mode_(0), ti_(0.0), td_(-1.0),
                  Kp_(0.0),
                  tau_(0.0),
                  tdByDt_(1.0),
                  dtByTi_(1.0),
                  e_(0.0),
                  e1_(0.0),
                  e2_(0.0)
{
}

VClamp::~VClamp()
{
    ;
}

void VClamp::setCommand(double value)
{
    cmdIn_ = value;
}

double VClamp::getCommand() const
{
    return cmdIn_;
}

void VClamp::setTi(double value)
{
    ti_ = value;
}

double VClamp::getTi() const
{
    return ti_;
}
void VClamp::setTd(double value)
{
    td_ = value;
}

double VClamp::getTd() const
{
    return td_;
}

void VClamp::setTau(double value)
{
    tau_ = value;
}

double VClamp::getTau() const
{
    return tau_;
}

void VClamp::setGain(double value)
{
    Kp_ = value;
}

double VClamp::getGain() const
{
    return Kp_;
}

double VClamp::getCurrent() const
{
    return current_;
}

void VClamp::setVin(double value)
{
    vIn_ = value;
}

double VClamp::getVin() const
{
    return vIn_;
}


void VClamp::setMode(unsigned int mode)
{
    mode_ = mode;
}

unsigned int VClamp::getMode() const
{
    return mode_;
}

void VClamp::process(const Eref& e, ProcPtr p)
{
    assert(ti_ > 0);
    assert(td_ >= 0);
    assert(tau_ > 0);
    double dCmd = cmdIn_ - oldCmdIn_;
    command_ = cmdIn_ + dCmd * ( 1 - tauByDt_) + (command_ - cmdIn_ + dCmd * tauByDt_) * expt_;
    oldCmdIn_ = cmdIn_;
    e_ = command_ - vIn_;
    if (mode_ == 0){
        current_ +=  Kp_ * ((1 + dtByTi_ + tdByDt_) * e_ - ( 1 + 2 * tdByDt_) * e1_ + tdByDt_ * e2_);
        e2_ = e1_;
        e1_ = e_;
    } else if (mode_ == DERIVATIVE_ON_PV){ // Here the derivative error term is replaced by process variable
        current_ +=  Kp_ * ((1 + dtByTi_) * e_ - e1_ + tdByDt_ * ( vIn_ - 2 * v1_ + e2_));
        e2_ = v1_;
        v1_ = vIn_;
        e1_ = e_;
    } else if (mode_ == PROPORTIONAL_ON_PV){ // Here the proportional as well as the derivative error term is replaced by process variable
        current_ +=  Kp_ * (vIn_ - v1_ + dtByTi_ * e_ + tdByDt_ * ( vIn_ - 2 * v1_ + e2_));
        e2_ = v1_;
        v1_ = vIn_;
    }
    currentOut()->send(e, current_);
}

void VClamp::reinit(const Eref& e, ProcPtr p)
{

    vIn_ = 0.0;
    v1_ = 0;
    command_ = cmdIn_ = oldCmdIn_ = e_ = e1_ = e2_ = 0;
    if (ti_ == 0){
        ti_ = p->dt;
    }
    if (td_ < 0){
        td_ = 0.0;
    }
    if (tau_ == 0.0){
        tau_ = 5 * p->dt;
    }
    if (p->dt / tau_ > 1e-15){
        expt_ = exp(-p->dt/tau_);
    } else {
        expt_ = 1 - p->dt/tau_;
    }
    tauByDt_ = tau_ / p->dt;
    dtByTi_ = p->dt/ti_;
    tdByDt_ = td_ / p->dt;

    vector<Id> compartments;
    unsigned int numComp = e.element()->getNeighbors(compartments, currentOut());
    if (numComp > 0){
        double Cm = Field<double>::get(compartments[0], "Cm");
    	if (Kp_ == 0){
            Kp_ = Cm / p->dt;
    	}
		command_ = cmdIn_ = oldCmdIn_ = 
           	Field<double>::get(compartments[0], "initVm");
    }
}

//
// VClamp.cpp ends here
